lib/browser/api/touch-bar.js

```diff
@@ -0,0 +1,205 @@
+const {EventEmitter} = require('events')
+
+let nextItemID = 1
+
+class TouchBar extends EventEmitter {
+
+  // Bind a touch bar to a window
+  static _setOnWindow (touchBar, window) {
+    if (window._touchBar != null) {
+      window._touchBar._removeFromWindow(window)
+    }
+
+    if (touchBar == null) {
+      window._setTouchBarItems([])
+      return
+    }
+
+    if (Array.isArray(touchBar)) {
+      touchBar = new TouchBar(touchBar)
+    }
+    touchBar._addToWindow(window)
+  }
+
+  constructor (items) {
+    super()
+
+    if (!Array.isArray(items)) {
+      throw new Error('The items object provided has to be an array')
+    }
+
+    this.windowListeners = {}
+    this.items = {}
+    this.ordereredItems = []
+    const registerItem = (item) => {
+      this.items[item.id] = item
+      item.on('change', () => {
+        this.emit('change', item.id, item.type)
+      })
+      if (item.child instanceof TouchBar) {
+        item.child.ordereredItems.forEach(registerItem)
+      }
+    }
+    items.forEach((item) => {
+      this.ordereredItems.push(item)
+      if (!(item instanceof TouchBarItem)) {
+        throw new Error('Each item must be an instance of a TouchBarItem')
+      }
+      registerItem(item)
+    })
+  }
+
+  // Called by BrowserWindow.setTouchBar
+  _addToWindow (window) {
+    const {id} = window
+
+    // Already added to window
+    if (this.windowListeners.hasOwnProperty(id)) return
+
+    window._touchBar = this
+
+    const changeListener = (itemID) => {
+      window._refreshTouchBarItem(itemID)
+    }
+    this.on('change', changeListener)
+
+    const interactionListener = (event, itemID, details) => {
+      const item = this.items[itemID]
+      if (item != null && item.onInteraction != null) {
+        item.onInteraction(details)
+      }
+    }
+    window.on('-touch-bar-interaction', interactionListener)
+
+    const removeListeners = () => {
+      this.removeListener('change', changeListener)
+      window.removeListener('-touch-bar-interaction', interactionListener)
+      window.removeListener('closed', removeListeners)
+      window._touchBar = null
+      delete this.windowListeners[id]
+    }
+    window.once('closed', removeListeners)
+    this.windowListeners[id] = removeListeners
+
+    window._setTouchBarItems(this.ordereredItems)
+  }
+
+  // Called by BrowserWindow.setTouchBar
+  _removeFromWindow (window) {
+    const removeListeners = this.windowListeners[window.id]
+    if (removeListeners != null) removeListeners()
+  }
+}
+
+class TouchBarItem extends EventEmitter {
+  constructor (config) {
+    super()
+    this.id = `${nextItemID++}`
+  }
+
+  _addLiveProperty (name, initialValue) {
+    const privateName = `_${name}`
+    this[privateName] = initialValue
```

Comment:
Can't this just be

```js
let liveValue = initialValue
Object.defineProperty(this, name, {
  get: function () {
    return liveValue
  },
  set: function (value) {
    liveValue = value
    this.emit('change')
  },
})
```

Then we aren't exposing the private variable on the `this` scope